openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -0,0 +1,186 @@
+        u'Version of the data at the time of collection.',
+        blank=True,
+        max_length=255,
+    )
+    data_edit_timestamp = models.DateTimeField(
+        u'Edit timestamp of the data at the time of collection.',
+        blank=True,
+        null=True,
+    )
+    transformers_schema_version = models.CharField(
+        u'Representation of the schema version of the transformers used during collection.',
+        blank=False,
+        max_length=255,
+    )
+    block_structure_schema_version = models.CharField(
+        u'Version of the block structure schema at the time of collection.',
+        blank=False,
+        max_length=255,
+    )
+    data = models.FileField(
+        upload_to=_path_name,
+        storage=_bs_model_storage(),
+    )
+
+    def get_serialized_data(self):
+        """
+        Returns the collected data for this instance.
+        """
+        return self.data.read()
+
+    @classmethod
+    def get(cls, data_usage_key):
+        """
+        Returns the entry associated with the given data_usage_key.
+        Raises:
+             BlockStructureNotFound if an entry for data_usage_key is not found.
+        """
+        try:
+            return cls.objects.get(data_usage_key=data_usage_key)
+        except cls.DoesNotExist:
+            raise BlockStructureNotFound(data_usage_key)
+
+    @classmethod
+    def update_or_create(cls, serialized_data, data_usage_key, **kwargs):
+        """
+        Updates or creates the BlockStructureModel entry
+        for the given data_usage_key in the kwargs,
+        uploading serialized_data as the content data.
+        """
+        bs_model, created = cls.objects.update_or_create(defaults=kwargs, data_usage_key=data_usage_key)
+        bs_model.data.save('', ContentFile(serialized_data))
+        if not created:
+            cls._delete_old_files(data_usage_key)
+        return bs_model, created
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return u', '.join(
+            u'{}: {}'.format(field_name, unicode(getattr(self, field_name)))
+            for field_name in
+            [
+                u'data_usage_key',
+                u'data_version',
+                u'data_edit_timestamp',
+                u'transformers_schema_version',
+                u'block_structure_schema_version',
+            ]
+        )
+
+    @classmethod
+    def _delete_old_files(cls, data_usage_key, num_to_keep=None):
+        """
+        Deletes previous file versions for data_usage_key.
+        """
+        if not is_enabled(PRUNE_OLD_VERSIONS):
+            return
+
+        if num_to_keep is None:
+            num_to_keep = settings.BLOCK_STRUCTURES_SETTINGS.get('PRUNE_KEEP_COUNT', 5)
+
+        try:
+            storage = _bs_model_storage()
+
+            all_files = cls._get_all_files(data_usage_key)
+            all_files_by_date = sorted(all_files, key=storage.modified_time)
```

Comment:
As of now, the number of `data_usage_keys` will be equivalent to the number of courses in the system.  The reason we use the term `data_usage_keys` instead of `course_keys` is because the BS layer was initially designed to work for any block-tree type, if/when needed in the future.  

We expect there to be at most only `PRUNE_KEEP_COUNT` (default 5) number of "files" per course.